build: make develop branch build again

jochenseeber opened the pull request:
### Why?

- Tried to build develop branch, but did not compile
- Mostly because some dependencies required Go 1.23, but the project was set to use Go 1.19
- PR upgrades to Go 1.24, makes develop build again, fixes unit tests and CI test

### What this PR does?

- Fixes unit tests and CI tests
- Updated Go to 1.24.7
- Required some follow up changes to make everything run again
    - Updated golangci-lint to 2.5.0
    - Updated k8s modules to 0.34.1
    - Updated code generation to write to the correct location
    - Updated Nix configuration
- CI script changes:
    - Added own fio container as there was no arm64 support
    - Added a provider option to select k3s
    - Added a data directory option because tests were failing on `/tmp`
    - Made script run as non-root user with sudo permission
- Updated Nix configuration to use Go 1.24.7
- Fixed Go lint errors
- Added a Vagrantfile to run a development VM for non-Linux users

**Checklist:**
- [x] Fixes #415
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Has the change log section been updated? → No functional changes
- [x] Commit has unit tests → No new tests, fixes existing ones
- [x] Commit has integration tests → Fixes existing ones

jochenseeber committed:
chore(build): make develop branch build again

- Some dependencies required Go 1.23, but project was configured for Go 1.19
- Integration tests were broken
- Updated Go to 1.24.7
- Required some follow up changes to make everything run again
    - Updated golangci-lint to 2.5.0
    - Updated k8s modules to 0.34.1
    - Updated code generation to write to the correct location
    - Updated Nix configuration
- CI script changes:
    - Added own fio container as there was no arm64 support
    - Added a provider option to select k3s
    - Added a data directory option because tests were failing on `/tmp`
    - Made script run as non-root user with sudo permission
- Updated Nix configuration to use Go 1.24.7
- Fixed Go lint errors
- Added a Vagrantfile to run a development VM for non-Linux users

Signed-off-by: Jochen Seeber <jochen@seeber.me>

tiagolobocastro commented on shell.nix:
```diff
@@ -24,29 +25,45 @@ pkgs.mkShell {
+    
+    # Temp directories should not be in project directory to avoid issues if it's mounted remotely
+    export CGO_ENABLED=0
+    export GOCACHE="$HOME/nix/.go/.cache"
```
Maybe something like this?

```suggestion
    export GOCACHE="$HOME/.cache/openebs/lvm-localpv/.go/.cache"
```

Btw, what's the issue with having this in the workspace when "mounted remotely"?

jochenseeber replied:
I am developing on a Mac, and using a Vagrant VM running Debian for development (no LVM on Mac). The project directory is mounted using a remote filesystem (vboxsf). Somehow, Go and the CI tests to not work when putting binaries and temp files in the project folder, but work fine if I e.g. put them into the home directory.

So basically this allows non-Linux developers work on the project.

tiagolobocastro commented on ci/ci-test.sh:
```diff
@@ -307,6 +318,28 @@ while test $# -gt 0; do
+      if [ -z "$1" ]; then
+        needs_help "Missing argument for --datadir"
+      fi
+      # Convert to absolute path if not already absolute
```
I see conver already below, maybe also a [ -d "$1" ] to ensure it's a path?

tiagolobocastro reviewed (approved):
Awesome, this was long time needed
Btw did you figure out why it was broken? It used to compile for me and then stopped when last I tried

tiagolobocastro commented:
oh, please take a look at the lint failure @jochenseeber

mhkarimi1383 commented on .github/workflows/build_and_push.yml:
```diff
@@ -43,10 +43,10 @@ jobs:
         uses: actions/setup-go@v5
         with:
-          go-version: 1.19.9
+          go-version: 1.24.7
```
Can you make it configurable via a single file

For example a file called 

.go-version

That contains the version
Since this is a repeated thing

tiagolobocastro replied:
Yeah I was wondering about that. Also there's a lot of duplicated code across the workflows, though also not sure if it's possible to improve it on GHA.
Anyway that's an existing issue... though the file would already be an improvement

mhkarimi1383 replied:
Can we use workflow dispatch?
Using nix shell will be wonderful

tiagolobocastro replied:
Not sure because that would run on a separate job, not this one.
I have seen a cool way of doing these things with nix on the [postgres](https://github.com/PostgREST/postgrest/blob/ea2c64da541f908af89c7f3630fcdf660196de47/.github/workflows/linkcheck.yaml#L13) but I'd suggest not changing CI too much on this PR and rather make another one

jochenseeber replied:
This PR is already pretty big, so I would like to change as little as possible… I'll do another one once this one is done.

tiagolobocastro replied:
Yeah I agree, let's tackle it on a separate PR

mhkarimi1383 commented on .github/workflows/staging.yml:
```diff
@@ -91,10 +91,10 @@ jobs:
         uses: actions/checkout@v4
 
-      - name: Set up Go 1.19
+      - name: Set up Go 1.24
```
Some setups and make calls are repeated, would be great to be singletone

mhkarimi1383 commented on Vagrantfile:
```diff
@@ -0,0 +1,250 @@
+# Vagrantfile to spin up a development VM using UTM
```
Is this needed when we have nix shell and nixos shell?

I mean we have to maintain multiple environments, and the chance of breaking one of them later on is high

jochenseeber replied:
Using Vagrant allows developers working on non-Linux machines (e.g. me ;-) to work on the project. nix-shell does not work on my machine, and I also was unable to spin up a nixos VM.

mhkarimi1383 replied:
CC: @tiagolobocastro

tiagolobocastro replied:
Yeah not sure, we'd need someone who knows vagrant to maintain it working
CC @abhilashshetty04 @dsharma-dc

dsharma-dc replied:
no idea about vagrant at the moment.

mhkarimi1383 commented on nix/sources.json:
```diff
@@ -1,14 +1,14 @@
 {
     "nixpkgs": {
-        "branch": "release-23.11",
+        "branch": "release-25.05",
```
I think the go version is mentioned in the shell.nix file too

mhkarimi1383 commented:
In rawfile we are completely based in Nix Shell in our CI, and it is so much easier to manage things and also we maintain the local development environment and CI environment at the same time

And we have faster setups since we have cache ;)

I would recommend doing the same here :)

jochenseeber committed:
chore(build): Fixed review issues

- Updated workflows to use golangci-lint 2.5.0
- Updated cache paths
- Added act to be able to run Github actions locally to test
- Fixed some typos

Signed-off-by: Jochen Seeber <jochen@seeber.me>

jochenseeber committed:
Merge branch 'develop' into 415-make-develop-branch-build

tiagolobocastro commented on deploy/yamls/lvmvolume-crd.yaml:
```diff
@@ -108,6 +114,8 @@ spec:
             required:
             - capacity
             - ownerNodeID
+            - shared
+            - thinProvision
```
we shouldn't be changing the schema, this may be a breaking change?

jochenseeber replied:
Not sure what caused this. I updated the dependencies and regenerated the CRD files. `thinProvision` was added with https://github.com/openebs/lvm-localpv/pull/30 from 2021. Any suggestions on what to do?

mhkarimi1383 replied:
I think this is a breaking change, that how field are handled (required vs. optional) I think it requires additional annotations in the code to mark them as optional

tiagolobocastro replied:
```
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum=yes;no
	ThinProvision string `json:"thinProvision,omitempty"`
```

Looks like `Required` together with `omitempty` didn't quite work and ended up becoming optional but on newer version of the generator the behavior was modified?
Then I suggest simply removing the Required line from the go code

tiagolobocastro commented on shell.nix:
```diff
@@ -24,29 +25,45 @@ pkgs.mkShell {
+    
+    # Temp directories should not be in project directory to avoid issues if it's mounted remotely
+    export CGO_ENABLED=0
+    export GOCACHE="$HOME/.cache/openebs/lvm-localpv/go/cache"
```
a lot of entries below still using $HOME/nix
```patch
index 586d781..99fb4ad 100644
--- a/shell.nix
+++ b/shell.nix
@@ -34,19 +34,21 @@ pkgs.mkShell {
     unset GOARCH
     unset GOOS
     unset GOROOT
+
+    OPENEBS_CACHE="$HOME/.cache/openebs/lvm-localpv"
     
     # Temp directories should not be in project directory to avoid issues if it's mounted remotely
     export CGO_ENABLED=0
-    export GOCACHE="$HOME/nix/.go/.cache"
+    export GOCACHE="$OPENEBS_CACHE/.go/.cache"
     export GOENV=off
-    export GOMODCACHE="$HOME/nix/.go/.modcache"
-    export GOPATH="$HOME/nix/.go"
+    export GOMODCACHE="$OPENEBS_CACHE/.go/.modcache"
+    export GOPATH="$OPENEBS_CACHE/.go"
     export GOPROXY=direct
     export GOTELEMETRY="off"
-    export GOTMPDIR="$HOME/nix/.go/.tmp"
+    export GOTMPDIR="$OPENEBS_CACHE/.go/.tmp"
     export GOTOOLCHAIN=local
     export PATH="$GOPATH/bin:$PATH"
-    export TMPDIR="$HOME/nix/.tmp"
+    export TMPDIR="$OPENEBS_CACHE/.tmp"
 
     mkdir -p "$GOCACHE"
     mkdir -p "$GOMODCACHE"
@@ -56,14 +58,14 @@ pkgs.mkShell {
     if [ "$IN_NIX_SHELL" = "pure" ]; then
       # working sudo within a pure nix-shell
       for sudo in /run/wrappers/bin/sudo /usr/bin/sudo /usr/local/bin/sudo /sbin/sudo /bin/sudo; do
-        mkdir -p $HOME/nix/bins
-        ln -sf $sudo $HOME/nix/bins/sudo
-        export PATH=$HOME/nix/bins:$PATH
+        mkdir -p $OPENEBS_CACHE/bins
+        ln -sf $sudo $OPENEBS_CACHE/bins/sudo
+        export PATH=$OPENEBS_CACHE/bins:$PATH
         break
       done
     else
-      rm $HOME/nix/bins/sudo 2>/dev/null || :
-      rmdir $HOME/nix/bins 2>/dev/null || :
+      rm $OPENEBS_CACHE/bins/sudo 2>/dev/null || :
+      rmdir $OPENEBS_CACHE/bins 2>/dev/null || :
     fi
```

jochenseeber replied:
Sorry, messed up the PR. Is there a way for me to test the pull request workflow in my cloned repo, so I can test everything works before updating the PR?

tiagolobocastro replied:
hmm I guess you could raise a PR against your own repo?

jochenseeber committed:
chore(build): fixed workflow issues

Signed-off-by: Jochen Seeber <jochen@seeber.me>

jochenseeber committed:
chore(build): still trying to get CI to run

* Removed Vagrant configuration
* Added manual trigger to PR workflow to allow testing

Signed-off-by: Jochen Seeber <jochen@seeber.me>

jochenseeber commented:
Aaand another round… Basing the CI on a nix environment and having the same for development, testing and CI sounds more and more desirable.

Changes:
* The lint action hat no HOME env var, handled that
* Added `workflow_dispatch` trigger to the `pull_request` action. This allows starting the action manually before creating a PR
* Removed the `Vagrantfile`. If this requires discussion, it is better moved into a separate PR

mhkarimi1383 commented:
> Aaand another round… Basing the CI on a nix environment and having the same for development, testing and CI sounds more and more desirable.
> 
> Changes:
> 
>     * The lint action hat no HOME env var, handled that
> 
>     * Added `workflow_dispatch` trigger to the `pull_request` action. This allows starting the action manually before creating a PR
> 
>     * Removed the `Vagrantfile`. If this requires discussion, it is better moved into a separate PR

Nice

We should talk about this later to see how we handle non-Linux users, But I think that Nix Shell works on Windows and MacOS too

jochenseeber committed:
chore(build): removed incorrect required annotations in CRDs

Signed-off-by: Jochen Seeber <jochen@seeber.me>

jochenseeber commented:
Ok, I removed the incorrect `required` annotations from the CRDs. There was one more (`snapSize`) in LVMSnaphot.

tiagolobocastro commented on deploy/yamls/lvmsnapshot-crd.yaml:
```diff
@@ -62,6 +66,7 @@ spec:
                 type: string
             required:
             - ownerNodeID
+            - snapSize
```
this one is still here

jochenseeber replied:
Will fix. There seems to be a major issue with mounting the project dir in Vagrant. Maybe a good thing to not add it…

tiagolobocastro replied:
It may be due to some caching, I also had to disable that on the nixos vm.nix: <https://github.com/openebs/lvm-localpv/blob/f103ba5e3fa2521a3bc3e6e9c53608942b950b5d/vm.nix#L18>

tiagolobocastro commented on pkg/mgmt/snapshot/snapshot.go:
```diff
@@ -34,7 +34,7 @@ import (
 // isDeletionCandidate checks if a lvm snapshot is a deletion candidate.
 func (c *SnapController) isDeletionCandidate(snap *apis.LVMSnapshot) bool {
-	return snap.ObjectMeta.DeletionTimestamp != nil
+	return snap.DeletionTimestamp != nil
```
isn't the `DeletionTimestamp` in the metadata?

jochenseeber replied:
Not sure, I'm new to Go. After upgrading the linter, it said the `.ObjectMeta` was redundant.

Abhinandan-Purkait replied:
Yeah, iirc it's a part of metadata.

tiagolobocastro replied:
Looks like this is a go feature, field promotion

dsharma-dc commented:
@jochenseeber I'd like to mention that while this PR is under review, it'll be considered for merging after the upcoming release branch has been cut, because the change touches quite a few files and needs some soak time.

jochenseeber committed:
chore(build): removed another incorrect required annotations in CRDs

Signed-off-by: Jochen Seeber <jochen@seeber.me>

abhilashshetty04 commented on pkg/apis/openebs.io/lvm/v1alpha1/lvmvolume.go:
```diff
@@ -77,14 +77,12 @@ type VolumeInfo struct {
 	// Shared specifies whether the volume can be shared among multiple pods.
 	// If it is not set to "yes", then the LVM LocalPV Driver will not allow
 	// the volumes to be mounted by more than one pods.
-	// +kubebuilder:validation:Required
```
Did you run `make manifests` after removing this tag. `shared` is not removed from required list of lvmvolume crd.

jochenseeber replied:
Should be fine with latest commit (05de9745e8bda7d8e46e574ac9dc669604debfb5)

abhilashshetty04 commented:
> Ok, I removed the incorrect `required` annotations from the CRDs. There was one more (`snapSize`) in LVMSnaphot.

@jochenseeber , Hasn't reflected in CRD yet. You have removed from field i see. Please do `make manifests`.

jochenseeber commented:
> > Ok, I removed the incorrect `required` annotations from the CRDs. There was one more (`snapSize`) in LVMSnaphot.
> 
> @jochenseeber , Hasn't reflected in CRD yet. You have removed from field i see. Please do `make manifests`.

Should be fine with latest commit (05de9745e8bda7d8e46e574ac9dc669604debfb5) from 6 hrs ago.

rohan2794 reviewed (commented):
@jochenseeber Avoid making changes to auto-generated code — these will be overwritten.

jochenseeber commented:
> @jochenseeber Avoid making changes to auto-generated code — these will be overwritten.

I believe I just regenerated the code. Did I miss something?

tiagolobocastro commented:
> > @jochenseeber Avoid making changes to auto-generated code — these will be overwritten.
> 
> I believe I just regenerated the code. Did I miss something?

@rohan2794 I also see no changes, `make kubegen` is generating the code in this PR

rohan2794 commented:
> > @jochenseeber Avoid making changes to auto-generated code — these will be overwritten.
> 
> I believe I just regenerated the code. Did I miss something?

Got it, looks like everything’s good then! Thanks for confirming.

rohan2794 commented:
> > > @jochenseeber Avoid making changes to auto-generated code — these will be overwritten.
> > 
> > 
> > I believe I just regenerated the code. Did I miss something?
> 
> @rohan2794 I also see no changes, `make kubegen` is generating the code in this PR

yes, @tiagolobocastro  , `kubegen` is generating the code.

dsharma-dc commented:
@jochenseeber I see you mentioned here: https://github.com/openebs/lvm-localpv/pull/416#issuecomment-3382801228. that vagrant file is removed. But I still that on this PR.

tiagolobocastro commented:
Thanks @jochenseeber for this much needed update!

Btw I was thinking that we could perhaps move the Vagrant file to a common repo for these things, where we could include other helpers, such as for running a talos cluster for example. Though the downside might be one would have to clone the common repo in addition to this one...

Thoughts? @Abhinandan-Purkait @mhkarimi1383
